### DISCUSSION TITLE
Support git sync from multiple repositories

### DISCUSSION THREAD

**Discussion by nareshurs** [apache/airflow#19381] [General]
**Description**

Git sync feature with multiple git repositories

**Use case / motivation**

We have a use case where multiple tenants create DAGs and they dont work with common git repository. Its a security issue if we share one customer info with other

**What do you want to happen?**

Airflow should support git-sync of DAG's from multiple git repositories

**Are you willing to submit a PR?**

No

**Related Issues**
No

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by EliMor** [comment]
seems related to: https://github.com/apache/airflow/issues/11708

**Comment by Bash-** [comment]
This would be nice to have, we also would like to attach multiple git repositories with DAGs, any updates?

The Bitnami Airflow Helm Chart supports this feature: https://github.com/bitnami/charts/blob/master/bitnami/airflow/values.yaml#L791

**Comment by potiuk** [comment]
Is there a reason why having single repo with submodules (each of the submodule repo in separate subdirectory) would not work for your case ?

**Reply by Francesko90** [reply]
I think that submodule are really hard to manage instead of classic repositories. Git submodules can add complexity to your repository, as they require additional commands and workflows that might be unfamiliar to some team members. This can lead to confusion, errors, and delays, moreover when using git submodules, it can be challenging to manage the version of the submodule, as different submodules might have different version requirements. This can lead to compatibility issues and make it harder to maintain the repository.

In the git submodule, the reference to specific subproject tag is saved as commit hash, so it is really easy to change ID commit for error and don't understand the reference to the code.

Anyway, I think that is a fascinating feature

**Reply by potiuk** [reply]
The thing is that if you are using multiple repos then you already ARE in the realm of complexity that submodule represents. How do you handle dependencies between those? How do you handle reverting changes. 

I wholeheartedly agree submodule is difficult to handle and you should avoid giving it to the hands of "regular" developers. This hit me EVERY SINGLE TIME I saw submodules introduced (or introduced it) as part of regular workflow by developers. No exceptions. On average submodule approach is not "teachable" - simply because as a regular developer you hit problems with it every few months and you have no chance to develop skills and "muscle memory". As opposed to regular git commands which most people use multiple times a day and develop the skills (relatively) quickly even if the concept of git is initially difficult to grasp.

But I am absolutely NOT proposing that.  The case I described this is very different. I propose that you have a separate, umbrella repository where you (manually or automatically as you wish) keep links between multiple repos using "submodule" features. But your regular developers might not even be aware of that, because each of them will only use their own repository. You just agree branching conventions, agree to the process of managint it and have a dedicated team (and automation) that will keep the shared submodule in sync.

This is precisely what Jagex team has done and spoke about in their Airflow Summit talk https://www.youtube.com/watch?v=uA-8Lj1RNgw. "Managing DAGs at scale". 

Generally speaking - developing support for multiple repos will very quickly turn into solving the same problme submodule solves. It will just be "poors-man-version" of it and you will have to build a tooling similar to submodule to manage all this. There is a very good example of other approaches that started from "yeah it must be simple to manage multiple repos". And I used few of those:
* subrepo (https://github.com/ingydotnet/git-subrepo) 
* repo (https://android.googlesource.com/tools/repo) 

All of them end up with basically the same - complex workflows, tooling with about a million switches that you have absolutely no clu about, various trade-offs. While your case might be simple, Airflow is a platform and must support multiple cases, users and scenarios, so once we relase "multi-repo" support, even if it will start simple, very soon some users will come up with expectations that they should be able to this and that. In a slightly different way. And as end result we will have our own "submodule" implementation - but likely partiy in XML and partily in python scripts that will lack documentation and have a lot of edge-cases unhandled. 

Do you think Airlfow community can do better than the authors of git, subrepo, repo? I have serious doubts.

**Reply by Francesko90** [reply]
Thank @potiuk, never thought that the airflow community can be better than authors of git. I think only that this feature can be reduced the complexity, but it's just my opinion and I agree with you that a new feature as it can be introduce other expectations in the future. 
I'm not ask to implement it but only defending the idea.

Thanks for the feedback :)

**Comment by potiuk** [comment]
@nareshurs @Bash- @EliMor - could you please comment on that 'submodule' ? I am asking because i am not a fan of 'we could do it because.we can' but more of a fan of 'use what's already available' so i am tempted to 'implement' that one by describing how you can use submodules to achieve what you want rather than support multiple repos.

I believe it has all you need for that (and even more because it centralizes management of multiple versions of different repos and keeps chart configuration simpler. 

I would love to hear if submodule approach works for you and if not -  WHY it would not work ?

**Comment by potiuk** [comment]
Here is a nice video showing how Jagex has done it with submodules: https://youtu.be/a-4yRne3ba4 - I would still love to hear if this is something that prevents you from doing this @Bash- @EliMor ?

**Reply by dszakallas** [reply]
It seems to me that they are packaging the dags in the image itself, there's no git-sync involved here.

**Comment by potiuk** [comment]
Converted it into a discussion. I believe git submodule support is great for supporitng multiple Git repositories and can handle pretty much all the cases with those much better than trying to implement multiple git repo support - also there are some examples and videos (see above) which explain how to do it, so unless there is a reason why this would not work (which can be discussed here) I just converted it into discussion.

**Reply by EliMor** [reply]
We went kind of a different route. Maybe the submodule idea is easier.
On merge to main/master of one of our dag repos the code is merged into a parent repo behind the scenes as a part of ci/cd. 

I think the argument of adding the feature to the chart comes from convenience than anything else. 

Thanks as always.

**Reply by potiuk** [reply]
It's not only convenience, it's also correctness (which we prefer not to sacrifice for convenience). There is an important difference between single and multiple repos - consistency between repos.

The main thing you get by havig single repo in git-sync (submodule or not) is that you have atomic sync of code. GitSync will first sync all the files and then replaces symbolic link to the checked-out folder atomically. This means that at any point in time you have "constistent" set of dags and libraries the dag uses - coming from the same commit as the "umbrella" repo for submodule or "main" repo in your case.  This is rather important when you have common code and cross-repo imports. 

Doing that when you have multiple repos is not really possible because semantically you cannot link two different commits between HEAD branches of the different repos. First of all you cannot push to two repos at the same time, and even if you do - pushing and pulling changes might take longer or shorter depending on the size of the changes, so there is no way to make sure the HEADs of two branches in two repos are in-sync. You can at most specify branches to use for those repos - unless you want to manually specify hash commits for each repo - but that's not practical.

So If you have multiple repos you can easily land in a situation that one of the repos refreshed to latest version, but the other did not do it yet - and you can have various kinds of import errors, failures, data corruption and the like.

The "single-repo" is better because effectively links betweeen the specific versions of the code from different repos are kept in git.
In submodules those are hash-commits of all linked repos stored for each submodule (this is how submodule works).  In your case this is the hash commit of the merge you do to the main repo.

**Reply by dszakallas** [reply]
### Q1: Import resolution
I wonder how import resolution works in this case? E.g we have `team1_repo` and `team2_repo` repos, each with their own set of python modules, some of them dags in it.
`team1_repo`:
```
team1
   __init__.py
   dag.py
   utils.py
```

`team2_repo`:
```
team2
   __init__.py
   dag.py
   utils.py
```

This way we can import `utils` within team1_repo using `import team1.utils`.
However when using submodules the dag root actually has a `team1_repo` directory in it, so that the import has to be changed to  `import team1_repo.team1.utils` to be correct? 
I don't find it a nice solution if this is the case. It'd be much better to have exactly the desired modules copied next to each other, i.e having the following dags home:
```
team2
   __init__.py
   dag1.py
   utils.py
team1
   __init__.py
   dag1.py
   utils.py
```
In our case there's no shared code between teams. Each have their own non-overlapping namespace. Is this solvable at all with git submodules in a non-hacky way?

With multiple repos, this is relatively easy, e.g an additional `relativeMountPath` property can determine the mountPath under the dags dir. E.g `team1` can go under `team1`, etc.

### Q2: Ignore the crust?

Specifying `subPath` is good way to remove unneccessary fluff and also to determine the import path. E.g in a scenario where the Python sources are under a `src` directory (common for modern repo layouts) `subPath` can point to `src` and have the module at the dag root. It seems to me that git submodules force the flat repo layout as we cannot have multiple `subPath`s.

With multiple repos and adding the relative mount path property this too is relatively easy, e.g by  `src/team1` going under `team1` under the dags dir. 

Thinking further about the `subPath` optimization, we could also prevent checking out the accompanying files in the repo that are not needed in the airflow environment, i.e test files, CI scripts, readmes, by doing a sparse checkout. git-sync already has [support](https://github.com/kubernetes/git-sync#flags-which-control-how-git-runs) for it. In case of a single repo and `subPath` given, there's easy to derive automatically a file that works well in most cases (ignore everything, and add back everything under `subPath`). Once again I don't know how this would work with submodules.

**Reply by ampx-mg** [reply]
@dszakallas How have you eventually solved the Q1 problem? I would be curious how other people deal with this, since the best solution I could come up with was adjusting the PYTHONPATH and I am not exceptionally thrilled about this

**Reply by artiya4u** [reply]
I've solved Q1 using a wrapped callable function and appended the PYTHONPATH inside the function using the different module paths for each repos from Airflow variable and no import modules on dag level code, only import inside the function.

Example DAG:
```python
def call_do_something(): # the wrapped function
    os.chdir(Variable.get('team1_path')) # team1 might use relative path
    sys.path.append(Variable.get('team1_path')) # path to team1
    from utils import do_something # utils.py will load from 'team1_path'
    do_something('param1')

default_args = {
    'owner': 'team1',
    'start_date': pendulum.today('UTC'),
    'depends_on_past': False,
    'email_on_failure': False,
    'email_on_retry': False,
    'retries': 3,
    'retry_delay': timedelta(minutes=1),
}

# define the DAG
with DAG(
        dag_id='team1_do_something',
        default_args=default_args,
        description="Run team1 do_something",
        catchup=False,
        max_active_runs=1,
        schedule="30 * * * *",
        tags=['team1'],
) as dag:
    start_dummy = EmptyOperator(task_id='start')
    end_dummy = EmptyOperator(task_id='end')
    target_task = PythonOperator(
        task_id=f'team1_do_something',
        python_callable=call_do_something,
        dag=dag,
        execution_timeout=timedelta(minutes=20),
    )
    start_dummy >> target_task >> end_dummy
```

**Comment by siddu2020** [comment]
I am also in need for something similar, I am trying to use airflow for one of the services I create for my customers we are in a place where customers can bring their own dags and I want to have a separate repo for each of them. is this possible ?

**Reply by potiuk** [reply]
Again - use submodules. 

This is the swiss-army nife for all things where you want to bring together several repositories. It's built-in in Git. There is no need to build something "special" for Airflow when the submodules solution is right there in git - supported by git-sync.

**Reply by Francesko90** [reply]
@siddu2020  Another option is to replace the git-sync container with a minio/s3 client container that fetch dags from an s3/minio bucket and you can create a CI that update the dags in the bucket

**Comment by prm-dan** [comment]
Does git-sync with submodules work when ssh deploy keys are used?  I'm hitting issues the following error.

```
Failed to add the ECDSA host key for IP address '140.82.113.3' to the list of known hosts (/etc/git-secret/known_hosts).\r\nERROR: Repository not found.\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\nfatal: clone of ...
```

**Reply by anirudh-kilambi** [reply]
In case you're still not sure on whether it works, my team was able to get git sync to work with SSH keys as well. It isn't close to a perfect solution yet, as I'm using my own personal public/private key pair and a subset of my own known hosts file with the hosts needed. 

I pass the predefined known_hosts file in my k8s secret for ssh keys.

Once again, not a perfect solution but git-sync works as expected when dealing with nested submodules.

One thing I am still trying to figure out is how to provide the sparse-checkout file -- the git-sync-init container passes `file not found` errors and I'm not sure whether I need to mount an extra volume for it to properly work.
